### ISSUE TITLE
tools.pythonpath doesn't work with source function

### ISSUE THREAD

**Issue by tru** [conan-io/conan#607]
if I do something like:

``` py
def source(self):
  with self.pythonpath(self):
    import mytools
    mytools.dosomething()
```

This works fine when running `conan install` with `conan source` it fails.

**Comment by memsharded** [comment]
Yeah, might be related somewhat to #604. We'll have a look

**Comment by memsharded** [comment]
You might need to use the generator "env" in your recipe, if you want to use this feature. All conan commands that do not retrieve and compute the full state, i.e. the full dependency graph, need to obtain the state from somewhere, basically the files created by generators:

- txt: conanbuildinfo.txt stores the information for ``deps_cpp_info``, typically necessary for ``build`` command, but maybe also others
- env: conanenv.txt. stores the information for ``deps_env_info``, typically environment variables like PATH or PYTHONPATH, the latter need for reusing python code from other recipes.

We still have to fully document the python code reuse functionality, sorry for that.

**Comment by tru** [comment]
Was this ever fixed? I still can't get it to work:

```
ERROR: zlib/1.2.8@plex/stable: Error in source() method, line 20
	import plexconantool
	No module named plexconantool
```

My zlib conanfile contains the line `generators = "env"`

What am I missing?

**Comment by memsharded** [comment]
Yes, I added a test in https://github.com/conan-io/conan/pull/609, you can check the PR, with the latest lines executing the ``source`` command. It is necessary to add the ``with tools.pythonpath(self):``, as other methods. 

Maybe the issue could be that it requires to ``conan install`` the package dependencies first, so the ``conanenv.txt`` is generated with the PYTHONPATH to the dependencies. Otherwise, I have no idea what could be failing, if it is not that, please tell us, and we could arrange a chat or something to try to help.

**Comment by memsharded** [comment]
After reviewing this issue, I have realized that this is a chicken and egg problem: If reusing python code from other package via ``tools.pythonpath``, it is necessary to install the dependencies first. But running ``conan source <package ref>`` is interesting as a first step prior to installing/building packages concurrently.

It works fine in user space, because ``conan install`` over a project in user space do not build/source/package anything for the project, only its dependencies. So it generates a ``conanenv.txt`` file, that the new local ``conan source`` implementation by @abuszta (to be released in 0.15), loads and uses successfully.

The only way the ``conan source <package ref>`` could work in that way would be to compute the dependency graph (as in ``conan install``), and use it instead of reading information from files generated at previous conan install. It could be argued that the same should/could be done to ``conan build`` and a possible future local ``conan package`` command. 

I see several advantages, as not having to declare/use the generators, nor read them from disk in conan build/source/package commands. The main disadvantage would be that computing the full dependency graph can be slow, especially for large projects, and that affects users as @youcefl with very large dependency graphs.

Please @lasote feedback on this issue

**Comment by tru** [comment]
If the concern is to add on a lot of time before source is executed (computing the dependency graph) maybe that can be a option or some attribute on the `ConanFile` since it's only important to people that reuse python code in the `source` function.

**Comment by tru** [comment]
@memsharded and @lasote I just want to make sure you guys see my last comment in this issue ^

**Comment by memsharded** [comment]
Actually, the ``txt`` generator could be also necessary for using the ``conan build`` command.
I have to do some profiling again, too, maybe opt-out with command line argument could be enough, having huge deps graphs is not that common.

**Comment by memsharded** [comment]
Some speed measurements: MacBookPro, Win10. Project with 250 dependencies packages:

- Deep (D1->D2->....->Dn): 1,6 secs load recipes + 3 secs propagate info in graph = 4,6 secs
- Breadth (D1->D0, D2->D0..., Dn->D0): 7,5 secs load recipes + 0,15 secs propagate info = 7,6 secs

The higher cost for the latest option is the detection of cycles in the dependency graph, which could be optionally disabled, then the cost would be about 5 secs for such a project, i.e. a 5 secs delay for ``source`` and ``build`` commands.

The first part is loading the recipes, which takes like 6ms, to load each recipe from disk, and load as python module (full interpreter, parsing), which seems difficult to optimize. The second part of propagating info, not easy, but could be somewhat improved, but even a 50% reduction, would leave us with 3 seconds delays.

I like the approach, it is more consistent and complete. Being it optional, would mean than the current implementation of using the generated ``env`` and ``txt`` files, invoking explicitly those generators, have to be maintained, and at some point the functionality might not be exactly the same.

The ``source`` command might be less trouble, it is mostly used in CI to parallelize builds, and if run locally is mostly for debugging/package creation. It actually doesn't need the dependencies binaries, so not necessary to run ``install`` before. But the ``build`` command might be more used during development, though I'd say that not very critical.

**Comment by memsharded** [comment]
@tru, I would need some feedback here: what is your usage pattern for ``conan source``?
Is it:

```bash
$ git clone ...
$ conan export myuser/mychannel
$ conan source MyPackage/0.1@myuser/mychannel
# In parallel
$ conan install ... # or conan test_package?
```

If it is that way, then we might be able to make it work just by installing the deps of MyPackage (to make sure we get the generated files with the data for ``pythonpath``) as:
```bash
$ git clone ...
$ conan install . -g env -g txt
$ conan export myuser/mychannel
$ conan source MyPackage/0.1@myuser/mychannel
...
```
Just considering other alternatives, retrieving and building the full deps graphs at each command has other implications that make it more difficult than expected. 

Thanks!

**Comment by tru** [comment]
Yeah that's my use case. I will try to implement that workaround and see if that works!

**Comment by memsharded** [comment]
Great, thanks, I will make a review too, analyze all alternatives, re-write some tests, and update here

**Comment by memsharded** [comment]
Should be fixed, will be in next release 0.15
